Added --fix option to CLI, w/ a "fix" hook that can be optionally added to any rule.

srowhani opened the pull request:
Similar to the functionality of eslint's --fix target, you write a hook and attach it to the object returned by the detect method in a given rule. This is then called from index.js for each error detected. Only rewrites the files that have a queued change. Not much was changed to support this functionality.

Any suggestions or contributions are more than welcome.

Feature originally requested by @michaelBenin here:

https://github.com/sasstools/sass-lint/issues/664

DanPurdy commented:
First of all thanks for this, definitely great to get the ball rolling, we'll have a look at it. First look the things I can see missing are

* [Developer Certificate of origin](https://github.com/sasstools/sass-lint/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin)
* Tests
* Documentation (including updating the rule documentation to indicate that it's 'fixable')

As this is going to be on a rule by rule basis we'll need to check it's how we'd like to proceed with it.

Thanks!

Snugug commented:
@srowhani Please address @DanPurdy's concerns or we will wind up needing to close this PR. His concerns reflect the whole  core team's concerns.

Snugug commented on index.js:
```diff
@@ -118,10 +116,13 @@ sassLint.lintText = function (file, options, configPath) {
+  if (ast.tree && ast.tree.content && ast.tree.content.length > 0) {
     rules.forEach(function (rule) {
-      detects = rule.rule.detect(ast, rule);
+      detects = rule.rule.detect(ast.tree, rule);
```
Do not complect the fix code with the lint code. These two things should live separately.

Snugug commented on lib/rules/property-sort-order.js:
```diff
@@ -143,5 +140,43 @@ module.exports = {
 
     return result;
+  },
+  'fix': function (groot, parser) {
```
Do not comma-separate off of `module.exports` for `'fix'`. Create a separate fix function and use `module.exports.fix = fix`

srowhani commented:
@Snugug @DanPurdy 

I'm glad your team is interested in my contribution. I've addressed some concerns and will soon include your contract. If you'd like to see progress that closer suits the needs of your team, perhaps it would be better to make contributions to the PR as well. Feel free to take the drivers seat on this.

srowhani commented:
Signed-off-by: Seena Rowhani <seenarowhani95@gmail.com>
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
660 York Street, Suite 102,
San Francisco, CA 94110 USA

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

jonny-improbable commented:
@srowhani thank you so much for this - I've just inherited a misconfigured project which contained ~400 `property-sort-order` errors which were fixed in under a second by checking out your branch 💯

bbtfr commented:
I'd love to see this PR get merged, it will make `sass-lint` more powerful than `scss-lint`.
Do we have a plan when we can merge it?

DanPurdy commented:
HI @bbtfr There are a few steps I outlined above which need to be taken care of before we can merge this. Most notably tests. We've been focusing on getting a lot of work out for 1.8 recently but now that is out of the way i think we can start to look at this properly. 

There are a few other outstanding omissions and PR's feature requests that we need to focus on too such as disabling the linters via comments but i dont see why this can't all be resolved soon though

bbtfr commented:
@DanPurdy Good to know! Thanks for your explanation.

DanPurdy commented:
@srowhani why do you keep closing this?
